CONTRIBUTING.md

```diff
@@ -0,0 +1,900 @@
+for (var i = 0; i < 52; i++) {
+	var j = i + (Math.random() * 52 - i)|0
+	var tmp = a[i]
+	a[i] = a[j]
+	a[j] = tmp
+}
+```
+
+### Distracting comments
+
+Don't comment *everything*. It's distracting, pointless, and a waste of time. Only comment when you need to.
+
+```js
+// Extremely bad. Don't *ever* do this.
+
+// This function loops through each item in a list, calling `f` with each item
+// and their respective index. If the function `f` returns explicitly with
+// `false`, iteration stops. This function returns the original list for
+// convenience.
+function forEach(list, f) {
+	// Loop through each entry in the list with the loop index `i`
+	for (var i = 0; i < list.length; i++) {
+		// Call the function with the current item and index
+		if (f(list[i], i) === false) {
+			// If the function explicitly returns `false`, immediately stop
+			// iteration.
+			break
+		}
+	}
+	// Return the list, for convenience.
+	return list
+}
+```
+
+### Initial whitespace
+
+Start your comments with a space. It's more readable.
+
+```js
+// Good
+// This is a comment.
+
+// Bad
+//This is a comment.
+```
+
+### Grammar and spelling
+
+Use proper grammar and spelling in your comments. This shouldn't need explanation.
+
+## Whitespace
+
+Use vertical whitespace and indentation to your advantage. It helps with readability, and minifiers are astonishingly great at removing this. :stuck_out_tongue_winking_eye:
+
+```js
+// Good
+function iterate() {
+	var index = 0
+	var list = []
+	while (hasNext()) {
+		list.push(next(index))
+		index++
+	}
+
+	for (var i = 0; i < list.length; i++) {
+		read(list[i])
+	}
+}
+
+// Bad
+function iterate() {
+	var index = 0
+	var list = []
+	while (hasNext()) list.push(next(index++))
+	for (var i = 0; i < list.length; i++) read(list[i])
+}
+
+// Even worse
+function iterate() {
+	var list = []
+	for (var index = 0; hasNext(); list.push(next(index++))) {}
+	for (var i = 0; i < list.length; i++) read(list[i])
+}
+```
+
+### Trailing whitespace
+
+Please don't leave trailing whitespace anywhere. It makes diffs harder to read.
+
+Blank lines should have no indentation or spaces. It should be empty other than a line ending.
+
+All files should end with a trailing line break. Some editors add this on their own, so it will result in unnecessary lines in later diffs if you don't.
+
+### Indentation and vertical whitespace
+
+Indent with hard tabs. Each one counts for 4 spaces.
+
+Never mix tabs and spaces. Don't use smart tabs.
+
+### Excessive whitespace
+
+Keep whitespace within reason. Limit vertical whitespace to no more than 2 consecutive blank lines, and don't start or end blocks with plain whitespace. Don't use more than one character of horizontal whitespace beyond indentation.
+
+```js
+// Bad
+function iterate() {
+	var index = 0
+	var list = []
+
+	while (hasNext()) {
+		list.push(next(index))
+		index++
+	}
+
+	// Too many empty lines
+
+
+
+	for (var i = 0; i < list.length; i++) {
+		read(list[i])
+	}
+}
```

Comment:
Could add suggestion. "If you're finding yourself resorting to multiple line breaks to visually separate logic, consider refactoring the code into smaller functions"